In [99]:
import pandas as pd

In [100]:
df = pd.read_csv('bank_scrooge.csv')

In [101]:
df.head()

USERID  score       city gender   age  equity     balance  products  \
0  183012  850.0    Рыбинск      Ж  25.0       1    59214.82         2   
1  146556  861.0    Рыбинск      Ж  37.0       5   850594.33         3   
2  120722  892.0    Рыбинск      Ж  30.0       0         NaN         1   
3  225363  866.0  Ярославль      Ж  51.0       5  1524746.26         2   
4  157978  730.0  Ярославль      М  34.0       5      174.00         1   

   credit_card  last_activity  EST_SALARY  churn  
0            0              1    75719.14      1  
1            1              0    86621.77      0  
2            1              1   107683.34      0  
3            0              1   174423.53      1  
4            1              0    67353.16      1

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   USERID         10000 non-null  int64  
 1   score          10000 non-null  float64
 2   city           10000 non-null  object 
 3   gender         10000 non-null  object 
 4   age            9974 non-null   float64
 5   equity         10000 non-null  int64  
 6   balance        7705 non-null   float64
 7   products       10000 non-null  int64  
 8   credit_card    10000 non-null  int64  
 9   last_activity  10000 non-null  int64  
 10  EST_SALARY     10000 non-null  float64
 11  churn          10000 non-null  int64  
dtypes: float64(4), int64(6), object(2)
memory usage: 937.6+ KB


In [103]:
# посмотрим min, max, среднее и тд.
df.describe().T

count           mean           std      min        25%  \
USERID         10000.0  171814.712600  3.370824e+04  94561.0  142810.25   
score          10000.0     848.699400  6.544852e+01    642.0     802.00   
age             9974.0      42.734409  1.217997e+01     18.0      33.00   
equity         10000.0       2.627600  1.980836e+00      0.0       0.00   
balance         7705.0  827794.306510  1.980614e+06      0.0  295554.16   
products       10000.0       1.870100  7.926467e-01      0.0       1.00   
credit_card    10000.0       0.680400  4.663449e-01      0.0       0.00   
last_activity  10000.0       0.523500  4.994724e-01      0.0       0.00   
EST_SALARY     10000.0  147866.885544  1.393885e+05   2546.3   75251.90   
churn          10000.0       0.182000  3.858638e-01      0.0       0.00   

                      50%          75%           max  
USERID         172728.000  201261.7500  2.291450e+05  
score             853.000     900.0000  1.000000e+03  
age                40.000      51.0000  8.600000e+01  
equity              3.000       4.0000  9.000000e+00  
balance        524272.200  980705.8500  1.191136e+08  
products            2.000       2.0000  5.000000e+00  
credit_card         1.000       1.0000  1.000000e+00  
last_activity       1.000       1.0000  1.000000e+00  
EST_SALARY     119658.105  174500.5425  1.395064e+06  
churn               0.000       0.0000  1.000000e+00

Переименуем столбцы

In [104]:
df.rename(columns={'USERID': 'user_id', 'EST_SALARY': 'est_salary'}, inplace = True)

Проверим пропущенные значения

In [105]:
df.isna().sum()

user_id             0
score               0
city                0
gender              0
age                26
equity              0
balance          2295
products            0
credit_card         0
last_activity       0
est_salary          0
churn               0
dtype: int64

Чтобы преобразовать значения столбца age к типу int, заменим пропущенные значения на 0. В дальнейшем будем учитывать, что минимальный возраст клиента - 18 лет.

In [106]:
df['age'] = df['age'].fillna(0)
df.isna().sum()# проверим результат

user_id             0
score               0
city                0
gender              0
age                 0
equity              0
balance          2295
products            0
credit_card         0
last_activity       0
est_salary          0
churn               0
dtype: int64

In [107]:
null_balance = df[df['balance'].isna()]
null_balance.head(10)

user_id  score       city gender   age  equity  balance  products  \
2    120722  892.0    Рыбинск      Ж  30.0       0      NaN         1   
9    133130  906.0  Ярославль      Ж  67.0       0      NaN         1   
10   148929  927.0     Ростов      М  52.0       0      NaN         1   
11   172184  921.0     Ростов      М  41.0       0      NaN         1   
19   127034  922.0    Рыбинск      Ж  53.0       0      NaN         1   
21   177415  896.0    Рыбинск      Ж  47.0       0      NaN         1   
30   155730  946.0  Ярославль      Ж  47.0       0      NaN         2   
34   205226  944.0  Ярославль      М  35.0       0      NaN         2   
39   173175  924.0    Рыбинск      Ж  51.0       0      NaN         1   
43   228585  956.0     Ростов      М  41.0       0      NaN         1   

    credit_card  last_activity  est_salary  churn  
2             1              1   107683.34      0  
9             0              1   238055.53      0  
10            1              1   196820.07      0  
11            1              1   217469.48      0  
19            0              0   147094.82      0  
21            0              0   225060.35      0  
30            1              1   660015.85      0  
34            1              1   239843.39      0  
39            1              1   248844.47      0  
43            1              0   115117.10      0

In [108]:
null_balance['equity'].value_counts()

0    2180
1     114
3       1
Name: equity, dtype: int64

Наличие пропущенных значений в столбце balance не имеет прямой связи ни с наличием кредитной карты, ни с количеством продуктов, которыми пользуется клиент, ни с приблизительной оценкой собственности клиента. Вероятно, наличие пропусков связано с тем, что данные клиенты не имеют депозитных, накопительных счетов или вкладов и пользуются другими продуктами и услугами банка: кредитные карты, страхование, банковские ячейки и др.

In [109]:
# проверим наличие явных дубликатов
print('Количество дубликатов в таблице:', df.duplicated().sum())

Количество дубликатов в таблице: 0


In [110]:
# проверим датафрейм на наличие неявных дубликатов
df.loc[df.duplicated(subset=['user_id'], keep=False)].sort_values('user_id')

user_id  score       city gender   age  equity      balance  products  \
1893   116540  883.0    Рыбинск      Ж  55.0       1    362756.49         3   
7694   116540  887.0  Ярославль      Ж  38.0       0          NaN         1   
7542   117943  880.0  Ярославль      Ж  40.0       0          NaN         1   
4866   117943  855.0    Рыбинск      Ж  32.0       6   1036832.93         4   
5896   120258  905.0  Ярославль      М  30.0       0          NaN         1   
...       ...    ...        ...    ...   ...     ...          ...       ...   
2597   226719  990.0  Ярославль      М  37.0       4  14648692.14         2   
8205   227795  840.0    Рыбинск      М  34.0       2    350768.03         1   
8497   227795  839.0  Ярославль      М  34.0       2    326593.14         2   
6457   228075  839.0    Рыбинск      М  39.0       5    507199.85         3   
1247   228075  932.0  Ярославль      М   0.0       5   7601719.20         2   

      credit_card  last_activity  est_salary  churn  
1893            0              1   175920.48      1  
7694            0              1   119247.61      0  
7542            1              0   137718.93      0  
4866            1              1   107792.71      1  
5896            1              1   146427.96      0  
...           ...            ...         ...    ...  
2597            0              0   934412.61      1  
8205            1              0   102036.14      1  
8497            1              0   103314.92      0  
6457            0              1    85195.80      0  
1247            1              1   408121.16      0  

[146 rows x 12 columns]

Здесь мы видим, что клиенты с одинаковыми id встречаются в разных городах. Вероятно, присвоение идентификатора пользователя в разных филиалах банка происходит независимо. Оставляем.

Изменим тип данных в столбцах score и age на int

In [111]:
df['score'] = df['score'].astype(int)
df['age'] = df['age'].astype(int)

Проверим на уникальность значения столбцов city и gender

In [112]:
df['city'].unique()

array(['Рыбинск', 'Ярославль', 'Ростов'], dtype=object)

In [113]:
df['gender'].unique()

array(['Ж', 'М'], dtype=object)

Добавим столбцы с наименованиями городов и по гендерному признаку

In [114]:
df['city_dummies'] = df['city']
df['gender_dummies'] = df['gender']
df = pd.get_dummies(df, columns=['city_dummies', 'gender_dummies'], prefix='', prefix_sep='', dtype=int)


user_id  score       city gender  age  equity     balance  products  \
0    183012    850    Рыбинск      Ж   25       1    59214.82         2   
1    146556    861    Рыбинск      Ж   37       5   850594.33         3   
2    120722    892    Рыбинск      Ж   30       0         NaN         1   
3    225363    866  Ярославль      Ж   51       5  1524746.26         2   
4    157978    730  Ярославль      М   34       5      174.00         1   
5    202305    856    Рыбинск      М   56       4   863687.24         3   
6    177259    807  Ярославль      Ж   39       3   405042.44         3   
7    218868    825  Ярославль      Ж   38       4   458145.40         2   
8    211686    923  Ярославль      М   54       5  1206337.87         2   
9    133130    906  Ярославль      Ж   67       0         NaN         1   
10   148929    927     Ростов      М   52       0         NaN         1   
11   172184    921     Ростов      М   41       0         NaN         1   
12   132970    845  Ярославль      Ж   31       3   591928.52         2   
13   166357    837    Рыбинск      Ж   28       4   302249.22         2   
14   172138    815  Ярославль      М   35       5   547499.87         2   
15   120260    731    Рыбинск      М   42       3  1480548.47         3   
16   123335    829  Ярославль      М   45       5   507842.84         1   
17   214277    793  Ярославль      М   53       3   627034.50         2   
18   136651    876  Ярославль      М   40       5  2307947.36         2   
19   127034    922    Рыбинск      Ж   53       0         NaN         1   

    credit_card  last_activity  est_salary  churn  Ростов  Рыбинск  Ярославль  \
0             0              1    75719.14      1       0        1          0   
1             1              0    86621.77      0       0        1          0   
2             1              1   107683.34      0       0        1          0   
3             0              1   174423.53      1       0        0          1   
4             1              0    67353.16      1       0        0          1   
5             1              0   156619.80      0       0        1          0   
6             0              1   103838.32      0       0        0          1   
7             1              1    68085.48      0       0        0          1   
8             1              0   155371.79      0       0        0          1   
9             0              1   238055.53      0       0        0          1   
10            1              1   196820.07      0       1        0          0   
11            1              1   217469.48      0       1        0          0   
12            1              1   112697.16      1       0        0          1   
13            0              1    30950.94      0       0        1          0   
14            1              1   105883.26      0       0        0          1   
15            1              0   160974.43      0       0        1          0   
16            1              1   169330.64      0       0        0          1   
17            1              0   183430.19      0       0        0          1   
18            0              0   287010.61      0       0        0          1   
19            0              0   147094.82      0       0        1          0   

    Ж  М  
0   1  0  
1   1  0  
2   1  0  
3   1  0  
4   0  1  
5   0  1  
6   1  0  
7   1  0  
8   0  1  
9   1  0  
10  0  1  
11  0  1  
12  1  0  
13  1  0  
14  0  1  
15  0  1  
16  0  1  
17  0  1  
18  0  1  
19  1  0

In [117]:
df.head(10)

user_id  score       city gender  age  equity     balance  products  \
0   183012    850    Рыбинск      Ж   25       1    59214.82         2   
1   146556    861    Рыбинск      Ж   37       5   850594.33         3   
2   120722    892    Рыбинск      Ж   30       0         NaN         1   
3   225363    866  Ярославль      Ж   51       5  1524746.26         2   
4   157978    730  Ярославль      М   34       5      174.00         1   
5   202305    856    Рыбинск      М   56       4   863687.24         3   
6   177259    807  Ярославль      Ж   39       3   405042.44         3   
7   218868    825  Ярославль      Ж   38       4   458145.40         2   
8   211686    923  Ярославль      М   54       5  1206337.87         2   
9   133130    906  Ярославль      Ж   67       0         NaN         1   

   credit_card  last_activity  est_salary  churn  Ростов  Рыбинск  Ярославль  \
0            0              1    75719.14      1       0        1          0   
1            1              0    86621.77      0       0        1          0   
2            1              1   107683.34      0       0        1          0   
3            0              1   174423.53      1       0        0          1   
4            1              0    67353.16      1       0        0          1   
5            1              0   156619.80      0       0        1          0   
6            0              1   103838.32      0       0        0          1   
7            1              1    68085.48      0       0        0          1   
8            1              0   155371.79      0       0        0          1   
9            0              1   238055.53      0       0        0          1   

   Ж  М  
0  1  0  
1  1  0  
2  1  0  
3  1  0  
4  0  1  
5  0  1  
6  1  0  
7  1  0  
8  0  1  
9  1  0

In [115]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   user_id        10000 non-null  int64  
 1   score          10000 non-null  int64  
 2   city           10000 non-null  object 
 3   gender         10000 non-null  object 
 4   age            10000 non-null  int64  
 5   equity         10000 non-null  int64  
 6   balance        7705 non-null   float64
 7   products       10000 non-null  int64  
 8   credit_card    10000 non-null  int64  
 9   last_activity  10000 non-null  int64  
 10  est_salary     10000 non-null  float64
 11  churn          10000 non-null  int64  
 12  Ростов         10000 non-null  int64  
 13  Рыбинск        10000 non-null  int64  
 14  Ярославль      10000 non-null  int64  
 15  Ж              10000 non-null  int64  
 16  М              10000 non-null  int64  
dtypes: float64(2), int64(13), object(2)
memory usage: 1

Вывод: перед нами данные о 10000 клиентах из 3 городов, мужчины и женщины в возрасте от 18 до 86 лет, с баллами
собственности от 642 до 1000, которые пользуются до 5 продуктами банка. Проведена проверка на наличие дубликатов, пропущенных значений.
В датафрейме встречаются строки с одинаковыми значениями id клиента. Так как эти пользователи из разных городов, передположим, что присвоение идентификатора пользователя в разных филиалах банка происходит независимо. Оставляем. Заменим пропущенные значения в столбце age на 0. В дальнейшем будем учитывать, что минимальный возраст клиента - 18 лет.
Наличие пропущенных значений в столбце balance не имеет прямой связи ни с наличием кредитной карты, ни с количеством продуктов, которыми пользуется клиент, ни с приблизительной оценкой собственности клиента. Вероятно, наличие пропусков связано с тем, что данные клиенты не имеют депозитных, накопительных счетов или вкладов и пользуются другими продуктами и услугами банка: кредитные карты, страхование, банковские ячейки и др.

## Исследовательский анализ

Рассчитаем доли мужчин и женщин среди клиентов банка

In [118]:
df['М'].mean()

0.5005

In [119]:
df['Ж'].mean()

0.4995

Рассчитаем доли мужчин и женщин среди лояльных и отточных клиентов

In [123]:
print('Доля мужчин среди лояльных клиентов:', round(100*df[df['churn']==0]['М'].mean(),2), '%')
print('Доля мужчин среди отточных клиентов:', round(100*df[df['churn']==1]['М'].mean(),2), '%')

Доля мужчин среди лояльных клиентов: 46.72 %
Доля мужчин среди отточных клиентов: 65.0 %


In [124]:
print('Доля женщин среди лояльных клиентов:', round(100*df[df['churn']==0]['Ж'].mean(),2), '%')
print('Доля женщин среди отточных клиентов:', round(100*df[df['churn']==1]['Ж'].mean(),2), '%')

Доля женщин среди лояльных клиентов: 53.28 %
Доля женщин среди отточных клиентов: 35.0 %


В целом в датафрейме отношение числа женщин и мужчин почти одинаково. Среди отточных клиентов больше мужчин, чем женщин. 65% клиентов, планирующих уйти из банка, - это мужчиныю

рассчитаем средний возраст клиента банка

In [125]:
df['age'].median()

40.0

рассчитаем средний возраст лояльного и отточного клиентов банка

In [130]:
print('Средний возраст лояльных клиентов:', round(df[df['churn']==0]['age'].median()))
print('Средний возраст отточных клиентов:', round(df[df['churn']==1]['age'].median()))

Средний возраст лояльных клиентов: 40
Средний возраст отточных клиентов: 39


Посмотрим, в каком городе отточных клиентов больше

In [137]:
print('Доля отточных клиентов в Ростове', round(100*df[df['churn']==1]['Ростов'].sum()/df['Ростов'].count(),2), '%')
print('Доля отточных клиентов в Рыбинске',round(100*df[df['churn']==1]['Рыбинск'].sum()/df['Рыбинск'].count(),2), '%')
print('Доля отточных клиентов в Ярославле',round(100*df[df['churn']==1]['Ярославль'].sum()/df['Ярославль'].count(),2), '%')

Доля отточных клиентов в Ростове 2.66 %
Доля отточных клиентов в Рыбинске 4.37 %
Доля отточных клиентов в Ярославле 11.17 %


Очень большой процент оттока можно отметить в Ярославле -11.17% от общего числа клиентов этого города, самый низкий процент оттока 2,66% - в Ростове